radix-engine/src/transaction/executor.rs

```diff
@@ -160,48 +160,45 @@ impl<'l, L: Ledger> TransactionExecutor<'l, L> {
         #[cfg(not(feature = "alloc"))]
         let now = std::time::Instant::now();
 
-        let signers = if let Some(Instruction::End { signers }) = transaction.instructions.last() {
-            // TODO: check all signer addresses are public key; eventually should be computed from signature.
-            signers.clone()
-        } else {
-            return Err(TransactionExecutionError::MissingEndInstruction);
-        };
+        let check_tx = transaction
+            .check()
+            .map_err(TransactionExecutionError::InvalidTransaction)?;
 
         let mut track = Track::new(
             self.ledger,
             self.current_epoch,
             sha256(self.nonce.to_string()),
-            signers,
+            check_tx.signers.clone(),
         );
         let mut proc = track.start_process(trace);
 
         let mut results = vec![];
         let mut success = true;
-        for inst in &transaction.instructions {
+        for inst in &check_tx.instructions {
             let res = match inst {
-                Instruction::DeclareTempBucket => {
+                CheckedInstruction::DeclareTempBucket => {
```

Comment:
Perhaps this should be `Instruction` and the current `Instruction` be called `UncheckedTransaction`

Reply:
`UncheckedTransaction` sounds a bit negative. I'd propose we keep them for now and if we really dislike it, we change them in a separate PR.

The background here is to introduce three phases of transaction processing:
1. Parsing
2. Static checking
3. Executing

`CheckedTransaction(CheckedInstruction)` is after phase 2.